rust/pact_verifier/src/pact_broker.rs

```diff
@@ -497,11 +508,140 @@ pub async fn fetch_pacts_from_broker(
               match pact_json {
                 Value::Object(ref map) => if map.contains_key("messages") {
                   match MessagePact::from_json(&href, &pact_json) {
-                    Ok(pact) => Ok((Box::new(pact) as Box<dyn Pact>, links)),
+                    Ok(pact) => Ok((Box::new(pact) as Box<dyn Pact>, None, links)),
+                    Err(err) => Err(PactBrokerError::ContentError(err))
+                  }
+                } else {
+                  Ok((Box::new(RequestResponsePact::from_json(&href, &pact_json)) as Box<dyn Pact>, None, links))
+                },
+                _ => Err(PactBrokerError::ContentError(format!("Link '{}' does not point to a valid pact file", href)))
+              }
+            },
+            Err(err) => Err(err)
+          }
+        })
+        .into_stream()
+        .collect()
+        .await;
+
+    Ok(results)
+}
+
+pub async fn fetch_pacts_dynamically_from_broker(
+  broker_url: String,
+  provider_name: String,
+  pending: bool,
+  include_wip_pacts_since: Option<String>,
+  provider_tags: Vec<String>,
+  consumer_version_selectors: Vec<ConsumerVersionSelector>,
+  auth: Option<HttpAuth>
+) -> Result<Vec<Result<(Box<dyn Pact>, Option<PactVerificationContext>, Vec<Link>), PactBrokerError>>, PactBrokerError> {
+    let mut hal_client = HALClient::with_url(broker_url.clone(), auth);
+    let template_values = hashmap!{ s!("provider") => provider_name.clone() };
+
+    hal_client = hal_client.navigate("pb:provider-pacts-for-verification", &template_values)
+    .await
+    .map_err(move |err| {
+      match err {
+        PactBrokerError::NotFound(_) =>
+        PactBrokerError::NotFound(
+          format!("No pacts for provider '{}' were found in the pact broker. URL: '{}'",
+          provider_name.clone(), broker_url.clone())),
+          _ => err
+        }
+      })?;
+
+    // Construct the Pacts for verification payload
+    let pacts_for_verification = PactsForVerificationRequest {
+      provider_version_tags: provider_tags,
+      include_wip_pacts_since: include_wip_pacts_since,
+      consumer_version_selectors: consumer_version_selectors,
+      include_pending_status: pending,
+    };
+    let request_body = serde_json::to_string(&pacts_for_verification).unwrap();
+
+    // Post the verification request
+    let response = match hal_client.find_link("self") {
+      Ok(link) => {
+        // TODO: make this an async stream to make error handling less shit
+        match hal_client.clone().post_json(hal_client.clone().parse_link_url(&link, &hashmap!{})?, request_body).await {
```

Comment:
If we split this into two statements and introduce a variable for the link URL, then we could get away without all the clones of the hal_client